official/utils/logging/logger.py

```diff
@@ -72,3 +84,106 @@ def log_metric(self, name, value, unit=None, global_step=None, extras=None):
       except (TypeError, ValueError) as e:
         tf.logging.warning("Failed to dump metric to log file: "
                            "name %s, value %s, error %s", name, value, e)
+
+  def log_run_info(self, model_name):
+    """Collect most of the TF runtime information for the local env.
+
+    The schema of the run info follows official/benchmark/datastore/schema.
+    """
+    run_info = {"model_name": model_name}
+    _collect_tensorflow_info(run_info)
+    _collect_environment_variable(run_info)
+    _collect_cpu_info(run_info)
+    _collect_gpu_info(run_info)
+    _collect_memory_info(run_info)
+
+    with tf.gfile.GFile(os.path.join(
+        self._logging_dir, _BENCHMARK_RUN_LOG_FILE_NAME), "w") as f:
+      try:
+        json.dump(run_info, f)
+        f.write("\n")
+      except (TypeError, ValueError) as e:
+        tf.logging.warning("Failed to dump benchmark run info to log file: %s",
+                           e)
+
+def _collect_tensorflow_info(run_info):
+  run_info["tensorflow_version"] = {
+      "version": tf.VERSION, "git_hash": tf.GIT_VERSION}
+
+def _collect_environment_variable(run_info):
+  run_info["environment_variable"] = {
+      k:v for k, v in os.environ.items() if k.startswith("TF_")}
+
+# The following code is mirrored from tensorflow/tools/test/system_info_lib
+# which is not exposed for import.
+def _collect_cpu_info(run_info):
+  cpu_info = {}
+
+  cpu_info["num_cores"] = multiprocessing.cpu_count()
+  # Gather num_cores_allowed
+  try:
+    with tf.gfile.GFile("/proc/self/status", "rb") as fh:
+      nc = re.search(r"(?m)^Cpus_allowed:\s*(.*)$", fh.read())
+    if nc:  # e.g. "ff" => 8, "fff" => 12
+      cpu_info["num_cores_allowed"] = (
+        bin(int(nc.group(1).replace(",", ""), 16)).count("1"))
+  except tf.OpError:
+    pass
+  finally:
+    if "num_cores_allowed" not in cpu_info:
+      cpu_info["num_cores_allowed"]= cpu_info["num_cores"]
+
+  info = cpuinfo.get_cpu_info()
+  cpu_info["cpu_info"] = info["brand"]
+  cpu_info["num_cores"] = info["count"]
+  cpu_info["mhz_per_cpu"] = info["hz_advertised_raw"][0] / 1.0e6
+  l2_cache_size = re.match(r"(\d+)", str(info.get("l2_cache_size", "")))
+  if l2_cache_size:
+    # If a value is returned, it"s in KB
+    cpu_info["cache_size"] = {"L2": int(l2_cache_size.group(0)) * 1024}
+
+  # Try to get the CPU governor
+  try:
+    cpu_governors = set([
+      tf.gfile.GFile(f, "r").readline().rstrip()
+      for f in glob.glob(
+          "/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor")
+    ])
+    if cpu_governors:
+      if len(cpu_governors) > 1:
+        cpu_info["cpu_governor"] = "mixed"
+      else:
+        cpu_info["cpu_governor"] = list(cpu_governors)[0]
+  except tf.OpError:
+    pass
```

Comment:
How frequently does something like this fail? My internal sirens are going off with all the hardcoded paths and regexes...

Reply:
I think it works for linux system mostly and should be quite stable. Windows will definitely break.